src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DirectiveAttributeCompletionItemProvider.cs

```diff
@@ -101,197 +112,289 @@ internal static ImmutableArray<RazorCompletionItem> GetAttributeCompletions(
-                            break;
-                        }
-                    }
+        // Use the mapping populated above to create completion items
+        return CreateCompletionItems(context, attributeCompletions);
+    }
+
+    private static ImmutableArray<(BoundAttributeDescriptor, string)> CollectIndexerDescriptors(
+        ImmutableArray<TagHelperDescriptor> descriptorsForTag,
+        DirectiveAttributeCompletionContext context)
+    {
+        if (context.InParameterName)
+        {
+            // No need to calculate the indexers When in a parameter name
+            return [];
+        }
+
+        using var allIndexers = new PooledArrayBuilder<(BoundAttributeDescriptor, string)>();
+
+        foreach (var descriptor in descriptorsForTag)
+        {
+            foreach (var attributeDescriptor in descriptor.BoundAttributes)
+            {
+                if (!attributeDescriptor.IsDirectiveAttribute)
+                {
+                    // We don't care about non-directive attributes
+                    continue;
                 }
 
                 if (!attributeDescriptor.IndexerNamePrefix.IsNullOrEmpty())
                 {
-                    TryAddCompletion(attributeDescriptor.IndexerNamePrefix + "...", attributeDescriptor, descriptor, razorCompletionOptions, selectedAttributeName, attributes, inSnippetContext, attributeCompletions);
+                    allIndexers.Add((attributeDescriptor, descriptor.TypeName));
                 }
             }
         }
 
+        return allIndexers.ToImmutableAndClear();
+    }
+
+    private static void AddAttributeNameCompletions(
+        TagHelperDescriptor descriptor,
+        BoundAttributeDescriptor attributeDescriptor,
+        DirectiveAttributeCompletionContext context,
+        Dictionary<string, (ImmutableArray<BoundAttributeDescriptionInfo>, ImmutableArray<RazorCommitCharacter>, RazorCompletionItemKind kind)> attributeCompletions)
+    {
+        if (!context.InAttributeName)
+        {
+            // Only add attribute name completions when in an attribute name context
+            return;
+        }
+
+        var isIndexer = context.SelectedAttributeName.EndsWith("...", StringComparison.Ordinal);
+        var descriptionInfo = BoundAttributeDescriptionInfo.From(attributeDescriptor, isIndexer, descriptor.TypeName);
+
+        if (!TryAddCompletion(attributeDescriptor.Name, descriptionInfo, descriptor, context, RazorCompletionItemKind.DirectiveAttribute, attributeCompletions) && attributeDescriptor.Parameters.Length > 0)
+        {
+            // This attribute has parameters and the base attribute name (@bind) is already satisfied. We need to check if there are any valid
+            // parameters left to be provided, if so, we need to still represent the base attribute name in the completion list.
+
+            foreach (var parameterDescriptor in attributeDescriptor.Parameters)
+            {
+                if (!context.ExistingAttributes.IsDefault
```

Comment:
Nit (and I'm guessing you didn't write this anyway): The IsDefault check could be outside the loop, though I'm not sure its needed at all with .Any

Reply:
It is needed. The ImmutableArray extension for Any doesn't check for the underlying array being null. I did it inside the loop as it seemed to fit nicer there, but I'll go ahead and move it outside